In [1]:
import pandas as pd
from pandas import json_normalize
import numpy as np
import json as js
import ast as ast
import re

In [2]:
games_org = pd.read_json('output_steam_games.json',lines=True)

In [210]:
games = games_org.copy()

In [3]:
list_rev = []
archivo1 = r'australian_user_reviews.json'
with open(archivo1, encoding='utf-8') as file:
    for line in file.readlines():
        list_rev.append(ast.literal_eval(line))

reviews_org = pd.DataFrame(list_rev)

In [122]:
reviews = reviews_org.copy()
reviews = reviews.explode('reviews',ignore_index=True)

In [4]:
list_items = []
archivo2 = r'australian_users_items.json'
with open(archivo2, encoding='utf-8') as file2:
    for line2 in file2.readlines():
        list_items.append(ast.literal_eval(line2))

items_org = pd.DataFrame(list_items)

In [ ]:
items = items_org.copy()
items = items.explode('items',ignore_index=True)

#### Revision de Games

In [211]:
games

publisher                                 genres  \
0                  None                                   None   
1                  None                                   None   
2                  None                                   None   
3                  None                                   None   
4                  None                                   None   
...                 ...                                    ...   
120440  Ghost_RUS Games  [Casual, Indie, Simulation, Strategy]   
120441           Sacada              [Casual, Indie, Strategy]   
120442     Laush Studio            [Indie, Racing, Simulation]   
120443         SIXNAILS                        [Casual, Indie]   
120444             None                                   None   

                        app_name                     title  \
0                           None                      None   
1                           None                      None   
2                           None                      None   
3                           None                      None   
4                           None                      None   
...                          ...                       ...   
120440            Colony On Mars            Colony On Mars   
120441  LOGistICAL: South Africa  LOGistICAL: South Africa   
120442             Russian Roads             Russian Roads   
120443       EXIT 2 - Directions       EXIT 2 - Directions   
120444               Maze Run VR                      None   

                                                      url release_date  \
0                                                    None         None   
1                                                    None         None   
2                                                    None         None   
3                                                    None         None   
4                                                    None         None   
...                                                   ...          ...   
120440  http://store.steampowered.com/app/773640/Colon...   2018-01-04   
120441  http://store.steampowered.com/app/733530/LOGis...   2018-01-04   
120442  http://store.steampowered.com/app/610660/Russi...   2018-01-04   
120443  http://store.steampowered.com/app/658870/EXIT_...   2017-09-02   
120444  http://store.steampowered.com/app/681550/Maze_...         None   

                                                     tags  \
0                                                    None   
1                                                    None   
2                                                    None   
3                                                    None   
4                                                    None   
...                                                   ...   
120440              [Strategy, Indie, Casual, Simulation]   
120441                          [Strategy, Indie, Casual]   
120442                        [Indie, Simulation, Racing]   
120443  [Indie, Casual, Puzzle, Singleplayer, Atmosphe...   
120444  [Early Access, Adventure, Indie, Action, Simul...   

                                              reviews_url  \
0                                                    None   
1                                                    None   
2                                                    None   
3                                                    None   
4                                                    None   
...                                                   ...   
120440  http://steamcommunity.com/app/773640/reviews/?...   
120441  http://steamcommunity.com/app/733530/reviews/?...   
120442  http://steamcommunity.com/app/610660/reviews/?...   
120443  http://steamcommunity.com/app/658870/reviews/?...   
120444  http://steamcommunity.com/app/681550/reviews/?...   

                                                    specs price  early_access  \
0                                       

In [212]:
games[(games['app_name'].isnull() & games['title'].isnull())]

publisher           genres app_name title  \
0          None             None     None  None   
1          None             None     None  None   
2          None             None     None  None   
3          None             None     None  None   
4          None             None     None  None   
...         ...              ...      ...   ...   
88307      None             None     None  None   
88308      None             None     None  None   
88309      None             None     None  None   
88384      None             None     None  None   
90890      None  [Action, Indie]     None  None   

                                               url release_date  \
0                                             None         None   
1                                             None         None   
2                                             None         None   
3                                             None         None   
4                                             None         None   
...                                            ...          ...   
88307                                         None         None   
88308                                         None         None   
88309                                         None         None   
88384               http://store.steampowered.com/         None   
90890  http://store.steampowered.com/app/317160/_/   2014-08-26   

                  tags                                        reviews_url  \
0                 None                                               None   
1                 None                                               None   
2                 None                                               None   
3                 None                                               None   
4                 None                                               None   
...                ...                                                ...   
88307             None                                               None   
88308             None                                               None   
88309             None                                               None   
88384             None                                               None   
90890  [Action, Indie]  http://steamcommunity.com/app/317160/reviews/?...   

                            specs  price  early_access        id developer  
0                            None   None           NaN       NaN      None  
1                            None   None           NaN       NaN      None  
2                            None   None           NaN       NaN      None  
3                            None   None           NaN       NaN      None  
4                            None   None           NaN       NaN      None  
...                           ...    ...           ...       ...       ...  
88307                        None   None           NaN       NaN      None  
88308                        None   None           NaN       NaN      None  
88309                        None   None           NaN       NaN      None  
88384                        None  19.99           0.0       NaN      None  
90890  [Single-player, Game demo]   None           0.0  317160.0      None  

[88312 rows x 13 columns]

120444 registros de games antes de borrar duplicados. Son 88309 registros duplicados y vacios. 3135 registros

In [213]:
games.drop(columns=['url','reviews_url'],inplace=True)

In [214]:
games.drop(games[games.duplicated(subset=['publisher', 'app_name', 'title', 'release_date', 'price', 'early_access', 'id',
       'developer'],keep=False)].index,inplace=True)

In [234]:
games.info()

<class 'pandas.core.frame.DataFrame'>
Index: 32133 entries, 88310 to 120444
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   genres        28850 non-null  object 
 1   app_name      32131 non-null  object 
 2   release_date  30066 non-null  object 
 3   tags          31970 non-null  object 
 4   specs         31463 non-null  object 
 5   price         30756 non-null  object 
 6   early_access  32133 non-null  float64
 7   id            32131 non-null  float64
 8   developer     28834 non-null  object 
dtypes: float64(2), object(7)
memory usage: 2.5+ MB


In [226]:
games.isnull().sum()

genres          3283
app_name           2
release_date    2067
tags             163
specs            670
price           1377
early_access       0
id                 2
developer       3299
dtype: int64

In [217]:
games.drop(columns=['publisher','title'],inplace=True)

In [232]:
games['price'].unique()

array([4.99, 0.0, 0.99, 2.99, 3.99, 9.99, 18.99, 29.99, None, 10.99,
       1.5899999999999999, 14.99, 1.99, 59.99, 8.99, 6.99, 7.99, 39.99,
       19.99, 7.49, 12.99, 5.99, 2.49, 15.99, 1.25, 24.99, 17.99, 61.99,
       3.49, 11.99, 13.99, 34.99, 74.76, 1.49, 32.99, 99.99, 14.95, 69.99,
       16.99, 79.99, 49.99, 5.0, 44.99, 13.98, 29.96, 119.99, 109.99,
       149.99, 771.71, 21.99, 89.99, 0.98, 139.92, 4.29, 64.99, 54.99,
       74.99, 0.89, 0.5, 299.99, 1.29, 3.0, 15.0, 5.49, 23.99, 49.0,
       20.99, 10.93, 1.3900000000000001, 36.99, 4.49, 2.0, 4.0, 9.0,
       234.99, 1.9500000000000002, 1.5, 199.0, 189.0, 6.66, 27.99, 10.49,
       129.99, 179.0, 26.99, 399.99, 31.99, 399.0, 20.0, 40.0, 3.33,
       199.99, 22.99, 320.0, 38.85, 71.7, 59.95, 995.0, 27.49, 3.39, 6.0,
       19.95, 499.99, 16.06, 4.68, 131.4, 44.98, 202.76, 1.0, 2.3,
       0.9500000000000001, 172.24, 249.99, 2.9699999999999998, 10.96,
       10.0, 30.0, 2.66, 6.48, 19.29, 11.15, 18.9, 2.89, 99.0, 87.94,
       5

In [220]:
games['price']=games['price'].replace('Starting at $499.00',499.00)
games['price']=games['price'].replace('Starting at $449.00',449.00)

In [221]:
mask = games['price'].apply(lambda x: isinstance(x, str))
games.loc[mask, 'price'] = 0.00

In [229]:
games2=games.copy()

In [231]:
games2['price']=games2['price'].astype(float)

ValueError: setting an array element with a sequence.

In [189]:
games.isna().sum().sort_values(ascending= False)

developer       3299
genres          3283
release_date    2067
price           1377
specs            670
tags             163
app_name           2
id                 2
early_access       0
dtype: int64

#### Revision de Reviews

In [123]:
reviews

user_id                                           user_url  \
0      76561197970982479  http://steamcommunity.com/profiles/76561197970...   
1      76561197970982479  http://steamcommunity.com/profiles/76561197970...   
2      76561197970982479  http://steamcommunity.com/profiles/76561197970...   
3                js41637               http://steamcommunity.com/id/js41637   
4                js41637               http://steamcommunity.com/id/js41637   
...                  ...                                                ...   
59328  76561198312638244  http://steamcommunity.com/profiles/76561198312...   
59329  76561198312638244  http://steamcommunity.com/profiles/76561198312...   
59330        LydiaMorley           http://steamcommunity.com/id/LydiaMorley   
59331        LydiaMorley           http://steamcommunity.com/id/LydiaMorley   
59332        LydiaMorley           http://steamcommunity.com/id/LydiaMorley   

                                                 reviews  
0      {'funny': '', 'posted': 'Posted November 5, 20...  
1      {'funny': '', 'posted': 'Posted July 15, 2011....  
2      {'funny': '', 'posted': 'Posted April 21, 2011...  
3      {'funny': '', 'posted': 'Posted June 24, 2014....  
4      {'funny': '', 'posted': 'Posted September 8, 2...  
...                                                  ...  
59328  {'funny': '', 'posted': 'Posted July 10.', 'la...  
59329  {'funny': '', 'posted': 'Posted July 8.', 'las...  
59330  {'funny': '1 person found this review funny', ...  
59331  {'funny': '', 'posted': 'Posted July 20.', 'la...  
59332  {'funny': '', 'posted': 'Posted July 2.', 'las...  

[59333 rows x 3 columns]

In [124]:
reviews.drop(columns=['user_url'],inplace=True)

In [12]:
reviews.iloc[1,1]

{'funny': '',
 'posted': 'Posted July 15, 2011.',
 'last_edited': '',
 'item_id': '22200',
 'helpful': 'No ratings yet',
 'recommend': True,
 'review': "It's unique and worth a playthrough."}

In [125]:
reviews_expanded = pd.json_normalize(reviews['reviews'])
reviews_expanded=reviews_expanded.replace('',None)

In [126]:
reviews_expanded.head()

funny                     posted last_edited item_id  \
0  None   Posted November 5, 2011.        None    1250   
1  None      Posted July 15, 2011.        None   22200   
2  None     Posted April 21, 2011.        None   43110   
3  None      Posted June 24, 2014.        None  251610   
4  None  Posted September 8, 2013.        None  227300   

                                           helpful recommend  \
0                                   No ratings yet      True   
1                                   No ratings yet      True   
2                                   No ratings yet      True   
3  15 of 20 people (75%) found this review helpful      True   
4     0 of 1 people (0%) found this review helpful      True   

                                              review  
0  Simple yet with great replayability. In my opi...  
1               It's unique and worth a playthrough.  
2  Great atmosphere. The gunplay can be a bit chu...  
3  I know what you think when you see this title ...  
4  For a simple (it's actually not all that simpl...

In [128]:
reviews.drop(columns=['reviews'],inplace=True)

In [136]:
reviews.isnull().sum()

user_id       0
posted        0
item_id       0
helpful       0
recommend     0
review       30
dtype: int64

In [130]:
reviews.drop(columns=['funny','last_edited'],inplace=True)

In [132]:
#28 registros que tienen datos nulos
reviews[reviews['posted'].isnull()]

user_id posted item_id helpful recommend review
137                gdxsd    NaN     NaN     NaN       NaN    NaN
177    76561198094224872    NaN     NaN     NaN       NaN    NaN
2559   76561198021575394    NaN     NaN     NaN       NaN    NaN
10080            cmuir37    NaN     NaN     NaN       NaN    NaN
13767          Jaysteeny    NaN     NaN     NaN       NaN    NaN
15493             ML8989    NaN     NaN     NaN       NaN    NaN
19184  76561198079215291    NaN     NaN     NaN       NaN    NaN
20223  76561198079342142    NaN     NaN     NaN       NaN    NaN
25056  76561198061996985    NaN     NaN     NaN       NaN    NaN
26257  76561198108286351    NaN     NaN     NaN       NaN    NaN
26674     CallTripleZero    NaN     NaN     NaN       NaN    NaN
28444       Priceless612    NaN     NaN     NaN       NaN    NaN
28452      diabolical666    NaN     NaN     NaN       NaN    NaN
28975         DrScottnik    NaN     NaN     NaN       NaN    NaN
31955  76561198110176420    NaN     NaN     NaN       NaN    NaN
32819         jojoPL1987    NaN     NaN     NaN       NaN    NaN
33785        Kylee_Kylie    NaN     NaN     NaN       NaN    NaN
39279             _samie    NaN     NaN     NaN       NaN    NaN
39357   FifthExclamation    NaN     NaN     NaN       NaN    NaN
39420             sbenji    NaN     NaN     NaN       NaN    NaN
40661     SpatialWheel14    NaN     NaN     NaN       NaN    NaN
41777  76561198065238937    NaN     NaN     NaN       NaN    NaN
44384  76561198071075080    NaN     NaN     NaN       NaN    NaN
50718  76561198063489150    NaN     NaN     NaN       NaN    NaN
53323   thecommunistbear    NaN     NaN     NaN       NaN    NaN
55425  76561198085883839    NaN     NaN     NaN       NaN    NaN
57578  76561198100396147    NaN     NaN     NaN       NaN    NaN
57766  76561198102423654    NaN     NaN     NaN       NaN    NaN

In [135]:
reviews.drop(reviews[reviews['posted'].isnull()].index,inplace=True)

In [119]:
reviews

user_id                     posted item_id  \
0      76561197970982479   Posted November 5, 2011.    1250   
1      76561197970982479      Posted July 15, 2011.   22200   
2      76561197970982479     Posted April 21, 2011.   43110   
3                js41637      Posted June 24, 2014.  251610   
4                js41637  Posted September 8, 2013.  227300   
...                  ...                        ...     ...   
59328  76561198312638244            Posted July 10.      70   
59329  76561198312638244             Posted July 8.  362890   
59330        LydiaMorley             Posted July 3.  273110   
59331        LydiaMorley            Posted July 20.     730   
59332        LydiaMorley             Posted July 2.     440   

                                               helpful recommend  \
0                                       No ratings yet      True   
1                                       No ratings yet      True   
2                                       No ratings yet      True   
3      15 of 20 people (75%) found this review helpful      True   
4         0 of 1 people (0%) found this review helpful      True   
...                                                ...       ...   
59328                                   No ratings yet      True   
59329                                   No ratings yet      True   
59330    1 of 2 people (50%) found this review helpful      True   
59331                                   No ratings yet      True   
59332                                   No ratings yet      True   

                                                  review  
0      Simple yet with great replayability. In my opi...  
1                   It's unique and worth a playthrough.  
2      Great atmosphere. The gunplay can be a bit chu...  
3      I know what you think when you see this title ...  
4      For a simple (it's actually not all that simpl...  
...                                                  ...  
59328  a must have classic from steam definitely wort...  
59329  this game is a perfect remake of the original ...  
59330  had so much fun plaing this and collecting res...  
59331                                                 :D  
59332                                     so much fun :D  

[58401 rows x 6 columns]

In [105]:
reviews.info()

<class 'pandas.core.frame.DataFrame'>
Index: 58459 entries, 0 to 59332
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   user_id      58459 non-null  object
 1   funny        8010 non-null   object
 2   posted       58431 non-null  object
 3   last_edited  6037 non-null   object
 4   item_id      58431 non-null  object
 5   helpful      58431 non-null  object
 6   recommend    58431 non-null  object
 7   review       58401 non-null  object
dtypes: object(8)
memory usage: 4.0+ MB


In [103]:
#Se borraron 874 duplicados
reviews.drop(reviews[reviews.duplicated()].index,inplace=True)

In [145]:
reviews.drop(columns=['helpful'],inplace=True)

#### Revision de Items

In [20]:
items

user_id  items_count           steam_id  \
0        76561197970982479          277  76561197970982479   
1        76561197970982479          277  76561197970982479   
2        76561197970982479          277  76561197970982479   
3        76561197970982479          277  76561197970982479   
4        76561197970982479          277  76561197970982479   
...                    ...          ...                ...   
5170010  76561198329548331            7  76561198329548331   
5170011  76561198329548331            7  76561198329548331   
5170012  76561198329548331            7  76561198329548331   
5170013  76561198329548331            7  76561198329548331   
5170014  edward_tremethick            0  76561198331598578   

                                                  user_url  \
0        http://steamcommunity.com/profiles/76561197970...   
1        http://steamcommunity.com/profiles/76561197970...   
2        http://steamcommunity.com/profiles/76561197970...   
3        http://steamcommunity.com/profiles/76561197970...   
4        http://steamcommunity.com/profiles/76561197970...   
...                                                    ...   
5170010  http://steamcommunity.com/profiles/76561198329...   
5170011  http://steamcommunity.com/profiles/76561198329...   
5170012  http://steamcommunity.com/profiles/76561198329...   
5170013  http://steamcommunity.com/profiles/76561198329...   
5170014     http://steamcommunity.com/id/edward_tremethick   

                                                     items  
0        {'item_id': '10', 'item_name': 'Counter-Strike...  
1        {'item_id': '20', 'item_name': 'Team Fortress ...  
2        {'item_id': '30', 'item_name': 'Day of Defeat'...  
3        {'item_id': '40', 'item_name': 'Deathmatch Cla...  
4        {'item_id': '50', 'item_name': 'Half-Life: Opp...  
...                                                    ...  
5170010  {'item_id': '373330', 'item_name': 'All Is Dus...  
5170011  {'item_id': '388490', 'item_name': 'One Way To...  
5170012  {'item_id': '521570', 'item_name': 'You Have 1...  
5170013  {'item_id': '519140', 'item_name': 'Minds Eyes...  
5170014                                                NaN  

[5170015 rows x 5 columns]

In [21]:
items.drop(columns=['user_url'],inplace=True)

In [22]:
items.iloc[0,3]

{'item_id': '10',
 'item_name': 'Counter-Strike',
 'playtime_forever': 6,
 'playtime_2weeks': 0}

In [26]:
items_expanded = pd.json_normalize(items['items'])
items_expanded=items_expanded.replace('',None)
items = items.join(items_expanded)
items.drop(columns=['items'],inplace=True)

In [24]:
items_expanded

item_id                      item_name  playtime_forever  \
0            10                 Counter-Strike               6.0   
1            20          Team Fortress Classic               0.0   
2            30                  Day of Defeat               7.0   
3            40             Deathmatch Classic               0.0   
4            50      Half-Life: Opposing Force               0.0   
...         ...                            ...               ...   
5170010  373330                    All Is Dust               0.0   
5170011  388490  One Way To Die: Steam Edition               3.0   
5170012  521570          You Have 10 Seconds 2               4.0   
5170013  519140                     Minds Eyes               3.0   
5170014     NaN                            NaN               NaN   

         playtime_2weeks  
0                    0.0  
1                    0.0  
2                    0.0  
3                    0.0  
4                    0.0  
...                  ...  
5170010              0.0  
5170011              3.0  
5170012              4.0  
5170013              3.0  
5170014              NaN  

[5170015 rows x 4 columns]

In [25]:
items_expanded.isnull().sum()

item_id             16806
item_name           16806
playtime_forever    16806
playtime_2weeks     16806
dtype: int64

In [143]:
items.drop(columns=['steam_id'],inplace=True)

In [93]:
items.isnull().sum()

user_id             0
items_count         0
steam_id            0
item_id             0
item_name           0
playtime_forever    0
playtime_2weeks     0
dtype: int64

In [101]:
#Se borran los duplicados
items.drop(items[items.duplicated()].index,inplace=True)

In [43]:
items['playtime_forever'].describe()

count    5.153209e+06
mean     9.914951e+02
std      5.418204e+03
min      0.000000e+00
25%      0.000000e+00
50%      3.400000e+01
75%      3.550000e+02
max      6.427730e+05
Name: playtime_forever, dtype: float64

In [38]:
items['playtime_forever'].mean()

np.float64(991.4951407947941)

In [41]:
items['playtime_forever'].max()

np.float64(642773.0)

In [39]:
items['playtime_2weeks'].mean()

np.float64(9.10470698937303)

In [40]:
items['playtime_2weeks'].max()

np.float64(19967.0)

In [28]:
items.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5170015 entries, 0 to 5170014
Data columns (total 7 columns):
 #   Column            Dtype  
---  ------            -----  
 0   user_id           object 
 1   items_count       int64  
 2   steam_id          object 
 3   item_id           object 
 4   item_name         object 
 5   playtime_forever  float64
 6   playtime_2weeks   float64
dtypes: float64(2), int64(1), object(4)
memory usage: 276.1+ MB


In [29]:
items.isnull().sum()

user_id                 0
items_count             0
steam_id                0
item_id             16806
item_name           16806
playtime_forever    16806
playtime_2weeks     16806
dtype: int64

16.806 nulls de 5.170.015 registros (0.32%)

In [33]:
items.dropna(inplace=True)

##### Tabla items user

In [151]:
items_user = items_org.copy()

In [153]:
items_user

user_id  items_count           steam_id  \
0            76561197970982479          277  76561197970982479   
1                      js41637          888  76561198035864385   
2                    evcentric          137  76561198007712555   
3                   Riot-Punch          328  76561197963445855   
4                        doctr          541  76561198002099482   
...                        ...          ...                ...   
88305        76561198323066619           22  76561198323066619   
88306        76561198326700687          177  76561198326700687   
88307  XxLaughingJackClown77xX            0  76561198328759259   
88308        76561198329548331            7  76561198329548331   
88309        edward_tremethick            0  76561198331598578   

                                                user_url  \
0      http://steamcommunity.com/profiles/76561197970...   
1                   http://steamcommunity.com/id/js41637   
2                 http://steamcommunity.com/id/evcentric   
3                http://steamcommunity.com/id/Riot-Punch   
4                     http://steamcommunity.com/id/doctr   
...                                                  ...   
88305  http://steamcommunity.com/profiles/76561198323...   
88306  http://steamcommunity.com/profiles/76561198326...   
88307  http://steamcommunity.com/id/XxLaughingJackClo...   
88308  http://steamcommunity.com/profiles/76561198329...   
88309     http://steamcommunity.com/id/edward_tremethick   

                                                   items  
0      [{'item_id': '10', 'item_name': 'Counter-Strik...  
1      [{'item_id': '10', 'item_name': 'Counter-Strik...  
2      [{'item_id': '1200', 'item_name': 'Red Orchest...  
3      [{'item_id': '10', 'item_name': 'Counter-Strik...  
4      [{'item_id': '300', 'item_name': 'Day of Defea...  
...                                                  ...  
88305  [{'item_id': '413850', 'item_name': 'CS:GO Pla...  
88306  [{'item_id': '11020', 'item_name': 'TrackMania...  
88307                                                 []  
88308  [{'item_id': '304930', 'item_name': 'Unturned'...  
88309                                                 []  

[88310 rows x 5 columns]

In [155]:
items_user.drop(columns=['steam_id','user_url'],inplace=True)